### PR TITLE
Dynamic datasources support

### LINKED ISSUE #87: Update Readme Generator to use New Gomplate Inline Datasources
## what
* New version gomplate of gomplate supports datasources to be defined inline

## why
* We should use this to simplify invocation

## references
- <https://github.com/hairyhenderson/gomplate/issues/349>
- https://github.com/hairyhenderson/gomplate/pull/373

### PR DESCRIPTION
**Pull request by vadim-gleyf**
## what
* Now use url from arg `datasource` if it passed
* Regenerate readme

## why
* https://github.com/cloudposse/build-harness/issues/87

### Depend on [#55](https://github.com/cloudposse/packages/pull/55)
* Need to bump `packages` [version](https://github.com/cloudposse/build-harness/blob/master/modules/packages/Makefile#L2) when it merged 

resolve: #87

### CODE REVIEW

**Review by osterman — CHANGES_REQUESTED** [review]
The objective is to move all data source definitions into the readme template. That way, we can simply pass —file and —out and it works. No obligation on the caller to provide data sources.

File: templates/README.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 <!-- This file was automatically generated by the `build-harness`. Make all changes to `README.yaml` and run `make readme` to rebuild this file. -->
-
+<!-- 
+{{ defineDatasource "config" .Env.DATASOURCE }}
```
Comment (osterman) [inline templates/README.md]: Use {{- ... -}} instead of the HTML comments

File: modules/readme/Makefile [hunk 2]
```diff
@@ -3,6 +3,9 @@ export README_FILE ?= README.md
 export README_YAML ?= README.yaml
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
 export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
+export USED_DATASOURCE ?= $(if $(datasource),$(datasource),$(shell pwd)/$(README_YAML))
```
Comment (osterman) [inline modules/readme/Makefile]: What is this for?

**Review by osterman — CHANGES_REQUESTED** [review]

File: modules/readme/Makefile (hunk 2 above)
Comment (vadim-gleyf) [reply modules/readme/Makefile]: @osterman  For correct output 
`@echo "Generated $(README_FILE) from $(README_TEMPLATE_FILE) using data from $(USED_DATASOURCE)"`

File: templates/README.md [hunk 3]
```diff
@@ -1,5 +1,6 @@
 <!-- This file was automatically generated by the `build-harness`. Make all changes to `README.yaml` and run `make readme` to rebuild this file. -->
-
+{{ defineDatasource "config" .Env.DATASOURCE }}
```
Comment (osterman) [inline templates/README.md]: This is what I meant we could do https://github.com/cloudposse/geodesic/blob/master/rootfs/templates/kops/default.yaml#L57

File: modules/readme/Makefile [hunk 4]
```diff
@@ -3,6 +3,9 @@ export README_FILE ?= README.md
...
+export DATASOURCE ?= $(if $(datasource),$(datasource),$(README_YAML))
```
Comment (osterman) [inline modules/readme/Makefile]: where is lowercase `datasource` ever defined?

File: modules/readme/Makefile [hunk 5]
```diff
@@ -29,7 +32,5 @@ readme/lint:
 ## Create README.md by building it from README.yaml
 readme/build: $(README_DEPS)
 	@gomplate --file $(README_TEMPLATE_FILE) \
-		--out $(README_FILE) \
-		--datasource config=$(README_YAML) \
-		--datasource includes='file://$(shell pwd)/?type=text/plain'
-	@echo "Generated $(README_FILE) from $(README_TEMPLATE_FILE) using data from $(README_TEMPLATE_YAML)"
+		--out $(README_FILE) 
+	@echo "Generated $(README_FILE) from $(README_TEMPLATE_FILE) using data from $(USED_DATASOURCE)"
```
Comment (osterman) [inline modules/readme/Makefile]: What's wrong with using `README_TEMPLATE_YAML`? I don't understand why we need a new variable.

File: templates/README.md [hunk 6]
```diff
@@ -1,5 +1,6 @@
 <!-- This file was automatically generated by the `build-harness`. Make all changes to `README.yaml` and run `make readme` to rebuild this file. -->
-
+{{ defineDatasource "config" .Env.DATASOURCE | regexp.Replace ".*" ""}}
```
Comment (osterman) [inline templates/README.md]: add a space between the quote and closing brace. `"}}`

File: templates/README.md [hunk 7]
```diff
@@ -1,5 +1,6 @@
...
+{{ defineDatasource "includes" .Env.INCLUDES | regexp.Replace ".*" ""}}
```
Comment (osterman) [inline templates/README.md]: add a space between the quote and closing brace. `"}}`

File: modules/readme/Makefile (hunk 4 above)
Comment (vadim-gleyf) [reply modules/readme/Makefile]: @osterman for overriding default datasource (`README.yaml` in root folder)
image

File: modules/readme/Makefile (hunk 5 above)
Comment (vadim-gleyf) [reply modules/readme/Makefile]: @osterman due possible overriding, it is not necessary?

File: modules/readme/Makefile (hunk 4 above)
Comment (osterman) [reply modules/readme/Makefile]: Is there a way we could just get rid of these envs and define it in the template directly?

**Review by osterman — CHANGES_REQUESTED** [review]
Can we get rid of the envs for the template and just hard code something in the template? My hope is to make the template almost entirely self-contained.

File: modules/readme/Makefile (hunk 4 above)
Comment (vadim-gleyf) [reply modules/readme/Makefile]: @osterman now if no any env passed, behaviour will be as earlier, will be used `README.yaml` from root project folder. 
This env just for possibility to change datasource, it isn't use case?

Comment (vadim-gleyf) [reply modules/readme/Makefile]: @osterman move datasource definition from makefile, but gomplate doesn't support inline shell execution, 
`export INCLUDES ?= $(file://$(shell pwd)/?type=text/plain)`
still here

File: modules/readme/Makefile [hunk 8]
```diff
@@ -3,6 +3,8 @@ export README_FILE ?= README.md
 export README_YAML ?= README.yaml
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
 export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
+export USED_DATASOURCE ?= $(if $(datasource),$(datasource),$(shell pwd)/$(README_YAML))
```
Comment (osterman) [inline modules/readme/Makefile]: Where does is `$(datasource)` defined? I would expect to see `datasource =` somewhere.

Comment (osterman) [reply modules/readme/Makefile]: The user can just specify `make readme README_TEMPLATE_YAML=foobar.yaml` if they want to override, so unsure what `datasource` gives us.

**Review by osterman — CHANGES_REQUESTED** [review]
See comments

File: modules/readme/Makefile [hunk 9]
```diff
@@ -2,8 +2,7 @@ export README_LINT ?= $(TMP)/README.md
 export README_FILE ?= README.md
 export README_YAML ?= README.yaml
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
-export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
-export USED_DATASOURCE ?= $(if $(datasource),$(datasource),$(shell pwd)/$(README_YAML))
+export README_TEMPLATE_YAML := $(if $(findstring http,$(README_YAML)),$(README_YAML),$(BUILD_HARNESS_PATH)/templates/$(README_YAML))
 export INCLUDES ?= $(file://$(shell pwd)/?type=text/plain)
```
Comment (osterman) [inline modules/readme/Makefile]: Rename to `README_INCLUDES` so we don't overload this env.

**Review by osterman — CHANGES_REQUESTED** [review]
Rename `INCLUDES`

File: modules/readme/Makefile [hunk 10]
```diff
@@ -3,7 +3,7 @@ export README_FILE ?= README.md
 export README_YAML ?= README.yaml
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
 export README_TEMPLATE_YAML := $(if $(findstring http,$(README_YAML)),$(README_YAML),$(BUILD_HARNESS_PATH)/templates/$(README_YAML))
-export INCLUDES ?= $(file://$(shell pwd)/?type=text/plain)
+export FILES ?= $(file://$(shell pwd)/?type=text/plain)
```
Comment (osterman) [inline modules/readme/Makefile]: No, rename to `README_INCLUDES` - the objective is to "namespace" environment variables related to the readme generation so they don't conflict. Generic names like `INCLUDES` or `FILES` are likely to "overload" other envs.

**Review by osterman — CHANGES_REQUESTED** [review]
See comments

**Review by osterman — DISMISSED** [review]

File: modules/readme/Makefile [hunk 11]
```diff
@@ -2,7 +2,8 @@ export README_LINT ?= $(TMP)/README.md
 export README_FILE ?= README.md
 export README_YAML ?= README.yaml
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
-export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
+export README_TEMPLATE_YAML := $(if $(findstring http,$(README_YAML)),$(README_YAML),$(BUILD_HARNESS_PATH)/templates/$(README_YAML))
```
Comment (osterman) [inline modules/readme/Makefile:5]: Can you elaborate on why we have findstring here?

**Review by osterman — COMMENTED** [review]
Explain need for `findstring`

File: modules/readme/Makefile (hunk 11 above)
Comment (vadim-gleyf) [reply modules/readme/Makefile:5]: @osterman to check is datasource external or not

**Review by osterman — APPROVED** [review]

**Pull request merged**